app/controllers/contacts_controller.rb

```diff
@@ -30,30 +34,54 @@ def spotlight
   private
 
   def set_up_contacts
+    if params[:a_id].present?
+      @aspect = current_user.aspects.find(params[:a_id])
+      gon.preloads[:aspect] = AspectPresenter.new(@aspect).as_json
+    end
+    @contacts_size = current_user.contacts.size
+  end
+
+  def set_up_contacts_json
     type = params[:set].presence
-    type ||= "by_aspect" if params[:a_id].present?
+    if params[:a_id].present?
+      type ||= "by_aspect"
+      @aspect = current_user.aspects.find(params[:a_id])
+    end
     type ||= "receiving"
-
-    @contacts = contacts_by_type(type)
-    @contacts_size = @contacts.length
-    gon.preloads[:contacts] = @contacts.map {|c| ContactPresenter.new(c, current_user).full_hash_with_person }
+    contacts_by_type(type).paginate(page: params[:page], per_page: 25)
+                          .map {|c| ContactPresenter.new(c, current_user).full_hash_with_person }
   end
 
   def contacts_by_type(type)
-    case type
+    order = ["profiles.first_name ASC", "profiles.last_name ASC", "profiles.diaspora_handle ASC"]
+    contacts = case type
       when "all"
+        order.unshift "receiving DESC"
         current_user.contacts
       when "only_sharing"
         current_user.contacts.only_sharing
       when "receiving"
         current_user.contacts.receiving
       when "by_aspect"
-        @aspect = current_user.aspects.find(params[:a_id])
-        gon.preloads[:aspect] = AspectPresenter.new(@aspect).as_json
-        current_user.contacts
+        order.unshift "contact_id IS NOT NULL DESC"
+        contacts_by_aspect(@aspect.id)
       else
         raise ArgumentError, "unknown type #{type}"
       end
+    contacts.includes(person: :profile)
+            .order(order)
+  end
+
+  def contacts_by_aspect(aspect_id)
+    contacts = current_user.contacts.arel_table
+    aspect_memberships = AspectMembership.arel_table
+    current_user.contacts.joins(
+      contacts.outer_join(aspect_memberships).on(
+        aspect_memberships[:aspect_id].eq(aspect_id).and(
+          aspect_memberships[:contact_id].eq(contacts[:id])
+        )
```

Comment:
Jotting down some of my code review notes in case someone finds it useful in the future:

This query converts to

```
SELECT `contacts`.* 
FROM `contacts` LEFT OUTER JOIN `aspect_memberships` 
ON `aspect_memberships`.`aspect_id` = THE_ASPECT_ID_YOU_ARE_SEARCHING_FOR AND `aspect_memberships`.`contact_id` = `contacts`.`id` 
WHERE `contacts`.`user_id` = YOUR_USER_ID`
```

Thus in the resulting table if the contact was not in the aspect_id you specified, the contact_id column will be null